openfl/databases/persistent_db.py

```diff
@@ -0,0 +1,299 @@
+        self.conn = sqlite3.connect(db_file, check_same_thread=False)
+        self.cursor = self.conn.cursor()
+        self.mutex = Lock()
+        self._create_model_tensors_table()
+        self._create_task_results_table()
+        self._create_key_value_Store()
+
+    def _create_model_tensors_table(self) -> None:
+        """Create the database schema if it does not exist."""
+        self.cursor.execute("""
+            CREATE TABLE IF NOT EXISTS tensors (
+                id INTEGER PRIMARY KEY AUTOINCREMENT,
+                tensor_name TEXT,
+                origin TEXT,
+                round INTEGER,
+                report INTEGER,
+                tags TEXT,
+                nparray BLOB
+            )
+        """)
+        self.conn.commit()
+    
+    def _create_task_results_table(self) -> None:
+        """Creates a table for storing task results."""
+        create_table_query = """
+        CREATE TABLE IF NOT EXISTS task_results (
+            id INTEGER PRIMARY KEY AUTOINCREMENT,
+            collaborator_name TEXT NOT NULL,
+            round_number INTEGER NOT NULL,
+            task_name TEXT NOT NULL,
+            data_size INTEGER NOT NULL,
+            named_tensors BLOB NOT NULL
+        );
+        """
+        self.cursor.execute(create_table_query)
+        self.conn.commit()
+    
+    def _create_key_value_Store(self) -> None:
+        self.cursor.execute("""
+            CREATE TABLE IF NOT EXISTS key_value_store (
+                key TEXT PRIMARY KEY,
+                value REAL
+            )
+        """)
+        self.conn.commit()
+    
+    def init_task_results_table(self):
+        """
+        Creates a table for storing task results. Drops the table first if it already exists.
+        """
+        drop_table_query = "DROP TABLE IF EXISTS task_results"
+        self.cursor.execute(drop_table_query)
+        
+        create_table_query = """
+        CREATE TABLE IF NOT EXISTS task_results (
+            id INTEGER PRIMARY KEY AUTOINCREMENT,
+            collaborator_name TEXT NOT NULL,
+            round_number INTEGER NOT NULL,
+            task_name TEXT NOT NULL,
+            data_size INTEGER NOT NULL,
+            named_tensors BLOB NOT NULL
+        );
+        """
+        self.cursor.execute(create_table_query)
+        self.conn.commit()
+
+    def save_task_results(
+        self,
+        collaborator_name: str,
+        round_number: int,
+        task_name: str,
+        data_size: int,
+        serialized_tensors,
+    ):
+        """
+        Saves task results to the task_results table.
+
+        Args:
+            collaborator_name (str): Collaborator name.
+            round_number (int): Round number.
+            task_name (str): Task name.
+            data_size (int): Data size.
+            named_tensors: Named tensors (e.g., protobuf, stored as serialized binary).
+        """
+        serialized_blob = pickle.dumps(serialized_tensors)
```

Comment:
This is already protobuf-serialized, why pickle it in addition? Unless you assume a non-serialized object could be passed here in certain cases?